prinz_pmml/src/main/scala/pl/touk/nussknacker/prinz/pmml/model/PMMLModelInstance.scala

```diff
@@ -1,11 +1,48 @@
 package pl.touk.nussknacker.prinz.pmml.model
 
-import org.jpmml.evaluator.Evaluator
-import pl.touk.nussknacker.prinz.model.ModelInstance
+import org.dmg.pmml.FieldName
+import org.jpmml.evaluator.{Evaluator, EvaluatorUtil, FieldValue, PMMLException}
+import pl.touk.nussknacker.engine.util.SynchronousExecutionContext.ctx
+import pl.touk.nussknacker.prinz.model.{ModelInstance, ModelRunException}
+import pl.touk.nussknacker.prinz.pmml.model.VectorMultimapUtils.VectorMultimapAsRowset
 import pl.touk.nussknacker.prinz.util.collection.immutable.VectorMultimap
 
+import java.util.{Map => JMap}
+import scala.jdk.CollectionConverters.{mapAsJavaMapConverter, mapAsScalaMapConverter}
+import scala.concurrent.Future
+
 case class PMMLModelInstance(evaluator: Evaluator, override val model: PMMLModel)
   extends ModelInstance(model, PMMLSignatureProvider) {
 
-  override def run(inputMap: VectorMultimap[String, AnyRef]): ModelRunResult = ???
+  type PMMLArgs = JMap[FieldName, FieldValue]
+
+  override def run(inputMap: VectorMultimap[String, AnyRef]): ModelRunResult = Future {
+    try {
+      val resultSeq = inputMap.forEachRow(evaluateRow)
+      val results = collectOutputs(resultSeq).asJava
+      Right(results)
+    } catch {
+      case ex: PMMLException => {
+        Left(new ModelRunException(ex.toString))
+      }
+    }
+  }
+
+  def evaluateRow(row: Map[String, AnyRef]): Map[String, _] = {
+    val args = EvaluatorUtil.encodeKeys(row.asJava)
+    val results = evaluator.evaluate(args)
+    EvaluatorUtil.decodeAll(results).asScala.toMap
+  }
+
+  def collectOutputs(rows: IndexedSeq[Map[String, _]]): Map[String, _] = {
+    rows.take(2).size match {
+      case 0 => Map[String, Any]()
+      case 1 => rows.head
+      case 2 => {
+        val vecMap = VectorMultimap[String, Any]()
+        rows.foreach(r => r.foreach(t => vecMap.add(t._1, t._2)))
```

Comment:
Vectormultimap is immutable so this operataion of adding new values returns new modified map an dimo this will not work - use another way to build this map

Reply:
Good catch! I used the builder shipped with the class.